pkg/metrics/metrics.jsx

```diff
@@ -440,17 +468,65 @@ class CurrentMetrics extends React.Component {
         }
 
         // top 5 CPU and memory consuming systemd units
-        newState.topServicesCPU = n_biggest(this.cgroupCPUNames, this.samples[9], 5).map(
-            ([key, value, is_user, is_container]) => cgroupRow(key, Number(value / 10 / numCpu).toFixed(1), is_user, is_container) // usec/s → percent
+        const topServicesCPU = n_biggest(this.cgroupCPUNames, this.samples[9], 5);
+        newState.topServicesCPU = topServicesCPU.map(
+            ([key, value, is_user, is_container, userid]) => cgroupRow(key, Number(value / 10 / numCpu).toFixed(1), is_user, is_container, userid) // usec/s → percent
         );
 
-        newState.topServicesMemory = n_biggest(this.cgroupMemoryNames, this.samples[10], 5).map(
-            ([key, value, is_user, is_container]) => cgroupRow(key, cockpit.format_bytes(value), is_user, is_container)
+        const topServicesMemory = n_biggest(this.cgroupMemoryNames, this.samples[10], 5);
+        newState.topServicesMemory = topServicesMemory.map(
+            ([key, value, is_user, is_container, userid]) => cgroupRow(key, cockpit.format_bytes(value), is_user, is_container)
         );
 
+        const notMappedContainers = topServicesMemory.concat(topServicesCPU).filter(([key, value, is_user, is_container, userid]) => is_container && getCachedPodName(userid, key) === undefined);
+        if (notMappedContainers.length !== 0) {
+            this.update_podman_name_mapping(notMappedContainers);
+        }
         this.setState(newState);
     }
 
+    /**
+     * Look up the container names using podman ps for the given cgroups. When
+     * the logged in user has administrative access we use sudo to call podman
+     * ps for that user.
+     */
+    update_podman_name_mapping = (cgroups) => {
+        // New mapping state
+        const podNameMapping = {};
+
+        const promises = cgroups.map(([containerid, value, is_user, is_container, userid]) => {
+            if (!(userid in podNameMapping)) {
+                podNameMapping[userid] = {};
+            }
+            // Always initialize the cache for when we hit an error.
+            podNameMapping[userid][containerid] = null;
+
+            if ((userid === 0 && !superuser.allowed) && userid !== this.state.userid) {
+                return null;
+            }
+            return cockpit.spawn(["podman", "ps", "--format", "json"], { superuser: userid === 0 ? "required" : null }).then(result => {
```

Comment:
Please move the .then() to the next line, it's easy to miss. Please also drop the `{ return }` wrapper.